70_FLIPFluids/gpuRadixSort.h

```diff
@@ -0,0 +1,470 @@
+		    cmdbuf->dispatch(numWorkgroups, 1, 1);
+
+			if (i < numIterations - 1) {
+				SMemoryBarrier memBarrier;
+				memBarrier.srcStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+				memBarrier.srcAccessMask = ACCESS_FLAGS::SHADER_WRITE_BITS;
+				memBarrier.dstStageMask = PIPELINE_STAGE_FLAGS::COMPUTE_SHADER_BIT;
+				memBarrier.dstAccessMask = ACCESS_FLAGS::SHADER_READ_BITS;
+				cmdbuf->pipelineBarrier(E_DEPENDENCY_FLAGS::EDF_NONE, {.memBarriers = {&memBarrier, 1}});
+            }
+        }
+    }
+
+private:
+    bool updateBuffers(int numElements, int numWorkgroups, int dataTypeSize)
+    {
+		bool updated = false;
+        if (!tempDataBuffer || tempDataBuffer->getSize() != numElements * dataTypeSize)
+        {
+            video::IGPUBuffer::SCreationParams params = {};
+		    params.size = numElements * dataTypeSize;
+		    params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+            tempDataBuffer = m_device->createBuffer(std::move(params));
+
+		    video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = tempDataBuffer->getMemoryReqs();
+		    reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+
+		    auto bufMem = m_device->allocate(reqs, tempDataBuffer.get());
+			updated = true;
+        }
+		uint32_t bufSize = sizeof(uint32_t) * numIterations * NUM_SORT_BINS;
+		if (!globalHistogramsBuffer || globalHistogramsBuffer->getSize() != bufSize)
+		{
+			video::IGPUBuffer::SCreationParams params = {};
+		    params.size = bufSize;
+		    params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+            globalHistogramsBuffer = m_device->createBuffer(std::move(params));
+
+		    video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = globalHistogramsBuffer->getMemoryReqs();
+		    reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+
+		    auto bufMem = m_device->allocate(reqs, globalHistogramsBuffer.get());
+			updated = true;
+		}
+		bufSize = sizeof(uint32_t) * (numWorkgroups * NUM_SORT_BINS + 1);
+		if (!partitionHistogramBuffer || partitionHistogramBuffer->getSize() != bufSize)
+		{
+			video::IGPUBuffer::SCreationParams params = {};
+		    params.size = bufSize;
+		    params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+            partitionHistogramBuffer = m_device->createBuffer(std::move(params));
+
+		    video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = partitionHistogramBuffer->getMemoryReqs();
+		    reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+
+		    auto bufMem = m_device->allocate(reqs, partitionHistogramBuffer.get());
+			updated = true;
+		}
+
+		return updated;
+    }
+
+    struct SSortParams
+    {
+        uint32_t numElements;
+        uint32_t bitShift;
+		uint32_t numWorkgroups;
+		uint32_t numThreadsPerGroup;
+    };
+
+	const uint32_t numThreadsPerGroup = 32;		// subgroup size
```

Comment:
the range of min-max subgroup size is meant to be queried from the `IPhysicalDevice`, then explicitly set when creating your pipeline